Yoast/wordpress-seo
# Allow flushing of custom taxonomies if their base changes

## Context


* Registers an additional check to the `wpseo_permalink_structure_check` cron job to check for changes in custom taxonomy slugs and resets the indexables table accordingly.

## Summary


This PR can be summarized in the following changelog entry:

* Registers an additional check to the `wpseo_permalink_structure_check` cron job to check for changes in custom taxonomy slugs and resets the indexables table accordingly.

## Relevant technical choices:

* The method will always reset the `custom_taxonomy_slugs` option to ensure that newly registered custom taxonomies also get taken into account in future checks.

## Test instructions


**Please note: This PR only checks for changes in the taxonomy's slug and does not touch the term in any way.**

This PR can be tested by following these steps:

* Ensure you've checked out this branch and have built everything.
* Ensure you've installed the Yoast Test helper.
* Ensure you've enabled the custom Books and Movies post types in the test helper.
* Add a movie or book category/genre if none exist.
* To trigger the the upgrade routine, first downgrade the version in the test helper to something like `15.1-RC0` and then set it back again to the current version.
* Check your database and ensure the `custom_taxonomy_slugs` property exists for the `wpseo` key and isn't empty.
* In your terminal, navigate to your Docker environment and run `./wp.sh cron event list`. See that the `wpseo_permalink_structure_check` cron is registered and is set for ~24 hours.

### Testing the cron job and resetting the database option
* In your database, change the `custom_taxonomy_slugs` property under the `wpseo` key and to something different than the current value. **Please replace a character instead of adding one as this won't break serialization**
* Force the cron to run by running `./wp.sh cron event run wpseo_permalink_structure_check`. You should receive a success message.
* Check your database and ensure the `custom_taxonomy_slugs` property still exists for the `wpseo` key and see that it has been restored to the proper set of custom taxonomies (i.e. a serialized array of custom taxonomies and their slugs).
* Check the `wp_yoast_indexable` table and see that all permalinks related to the custom taxonomy you edited earlier on, has been reset.
* Log into your admin and re-run indexation or rerun indexation via the WP-CLI (`./wp.sh yoast index --reindex`). **Please note that you should not have your environment set to anything other than `production` to use the CLI command, due to a bug that currently exists in the code**
* Check the `wp_yoast_indexable` table and see that permalinks have returned for the specific taxonomy.
* Run the `./wp.sh cron event run wpseo_permalink_structure_check` command again.
* Check the `wp_yoast_indexable` table and see that permalinks remain untouched.

### Testing changes made to the registered taxonomy
* Ensure you have at least two movie genres registered. Assign both to a post.
* Go back to the Genre overview and click on the View link for one of the genres you created. You should see the assigned post.
* In your editor of choice, open the Yoast Test Helper's code (File: `src/taxonomies.php`) and scroll down to line 90.
* Edit the slug (third parameter) by changing one of the words or by replacing a single letter. Save your changes.
* Refresh the page you viewed for the genre (step 2). You should see that the permalink **hasn't** changed and still works.
* Run the `./wp.sh cron event run wpseo_permalink_structure_check` command.
* Check the `wp_yoast_indexable` table and see that all permalinks related to the custom taxonomy's genres, have been reset.
* Refresh the page you were viewing. You should get a 404 because the indexable no longer exists.
* Change the `yoast-test-movie` part of the URL in your browser, to be the same as the edit made in the test helper's code.
* You should see the genre again, with the new taxonomy slug in the URL.
* Check the `wp_yoast_indexable` table (filter on `permalink` with a `LIKE` with the value `%genre%`) and see that all permalinks related to the custom taxonomy and genre you edited earlier on, has been reset. **You should not see the second genre just yet**.
* Rerun indexation via the WP-CLI (`./wp.sh yoast index --reindex`). **Please note that you should not have your environment set to anything other than `production` to use the CLI command, due to a bug that currently exists in the code**
* Check the `wp_yoast_indexable` table (filter on `permalink` with a `LIKE` with the value `%genre%`) once more and see that all permalinks related to the genres, have been reset.

## Post-testing
* To rollback the changes made in the test helper, rollback the changes made in the test helper and run both the `./wp.sh cron event run wpseo_permalink_structure_check` and `./wp.sh yoast index --reindex` commands.

## UI changes

## Documentation

## Quality assurance


Fixes [P3-153]
